File: VB/MyDateEdit/MyVistaDataEditCalendar.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Reflection
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows.Forms
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Controls
Imports DevExpress.XtraEditors.Popup
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.XtraEditors.ViewInfo
Imports DevExpress.Utils.Drawing

Namespace WindowsFormsApplication2
	Public Class MyVistaDataEditCalendar
		Inherits VistaDateEditCalendar
		Public Sub New(ByVal item As RepositoryItemDateEdit, ByVal editDate As Object)
			MyBase.New(item, editDate)
		End Sub

		Protected Overrides Function CreatePainter() As DevExpress.XtraEditors.Drawing.DateEditPainter
			Return New MyVistaDateEditPainter(Me)
		End Function

		Protected Overrides Function CreateInfoArgs() As DevExpress.XtraEditors.ViewInfo.DateEditInfoArgs
			Return New MyVistaDateEditInfoArgs(calendar:=Me)
		End Function

		Friend Shadows ReadOnly Property CaptionRect() As Rectangle
			Get
				Return MyBase.CaptionRect
			End Get
		End Property

		Protected Overrides Sub OnPressedObjectChanged(ByVal prev As DevExpress.XtraEditors.Calendar.CalendarHitInfo, ByVal curr As DevExpress.XtraEditors.Calendar.CalendarHitInfo)
			MyBase.OnPressedObjectChanged(prev, curr)
			Dim vdi As MyVistaDateEditInfoArgs = TryCast(Calendars(0), MyVistaDateEditInfoArgs)
			Dim hitInfo As MyCalendarHitInfo = TryCast(curr, MyCalendarHitInfo)
			Invalidate(prev.Bounds)
			If curr.HitObject IsNot Nothing Then
				If hitInfo.InLeftYearButton Then
					vdi.LeftYearArrowInfo.State = ObjectState.Pressed
				End If
				If hitInfo.InRightYearButton Then
					vdi.RightYearArrowInfo.State = ObjectState.Pressed
				End If
			End If
			If curr Is Nothing Then
				Invalidate(Bounds)
			Else
				Invalidate(curr.Bounds)
			End If
		End Sub

		Protected Overrides Sub OnPressedObjectChanging(ByVal prev As DevExpress.XtraEditors.Calendar.CalendarHitInfo, ByVal curr As DevExpress.XtraEditors.Calendar.CalendarHitInfo)
			MyBase.OnPressedObjectChanging(prev, curr)
			Dim vdi As MyVistaDateEditInfoArgs = TryCast(Calendars(0), MyVistaDateEditInfoArgs)
			Dim hitInfo As MyCalendarHitInfo = TryCast(curr, MyCalendarHitInfo)
			If curr.HitObject IsNot Nothing Then
				vdi.LeftYearArrowInfo.State = If(hitInfo.InLeftYearButton, ObjectState.Hot, ObjectState.Normal)
				vdi.RightYearArrowInfo.State = If(hitInfo.InRightYearButton, ObjectState.Hot, ObjectState.Normal)
			End If
		End Sub

		Protected Overrides Sub OnHotObjectChanged(ByVal prevInfo As DevExpress.XtraEditors.Calendar.CalendarHitInfo, ByVal currInfo As DevExpress.XtraEditors.Calendar.CalendarHitInfo)
			MyBase.OnHotObjectChanged(prevInfo, currInfo)
			Invalidate(prevInfo.Bounds)
			Dim vdi As MyVistaDateEditInfoArgs = TryCast(Calendars(0), MyVistaDateEditInfoArgs)
			Dim hitInfo As MyCalendarHitInfo = TryCast(currInfo, MyCalendarHitInfo)
			vdi.OnAnimationChanged(prevInfo, currInfo)
			If currInfo IsNot Nothing AndAlso currInfo.HitObject IsNot Nothing Then
				If hitInfo.InLeftYearButton AndAlso vdi.LeftYearArrowInfo.State <> ObjectState.Pressed Then
					vdi.LeftYearArrowInfo.State = ObjectState.Hot
				End If
				If hitInfo.InRightYearButton AndAlso vdi.RightYearArrowInfo.State <> ObjectState.Pressed Then
					vdi.RightYearArrowInfo.State = ObjectState.Hot
				End If
			End If
		End Sub

		Protected Overrides Sub OnHotObjectChanging(ByVal prevInfo As DevExpress.XtraEditors.Calendar.CalendarHitInfo, ByVal currInfo As DevExpress.XtraEditors.Calendar.CalendarHitInfo)
			MyBase.OnHotObjectChanging(prevInfo, currInfo)
			Dim vdi As MyVistaDateEditInfoArgs = TryCast(Calendars(0), MyVistaDateEditInfoArgs)
			vdi.OnAnimationChanging(prevInfo, currInfo)
			vdi.RightYearArrowInfo.State = ObjectState.Normal
			vdi.LeftYearArrowInfo.State = ObjectState.Normal
		End Sub

		Protected Overrides Sub ProcessClick(ByVal hit As DevExpress.XtraEditors.Calendar.CalendarHitInfo)
			MyBase.ProcessClick(hit)
			If hit.HitObject IsNot Nothing Then
				SwitchType = SwitchStateType.SwitchToRight
				OnStateChanging(View, View, False, True)
				Dim vdi As MyVistaDateEditInfoArgs = TryCast(Calendars(0), MyVistaDateEditInfoArgs)
				Dim hitInfo As MyCalendarHitInfo = TryCast(hit, MyCalendarHitInfo)
				If hitInfo.InRightYearButton Then
					DateTime = DateTime.AddYears(1)
				End If
				If hitInfo.InLeftYearButton Then
					DateTime = DateTime.AddYears(-1)
				End If
				OnStateChanged(View, View, False, True)
			End If
		End Sub

		Friend Shadows Property PrepareAnimation() As Boolean
			Get
				Return MyBase.PrepareAnimation
			End Get
			Set(ByVal value As Boolean)
				MyBase.PrepareAnimation = value
			End Set
		End Property
	End Class
End Namespace

File: CS/MyDateEdit/MyVistaDataEditCalendar.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Controls;
using DevExpress.XtraEditors.Popup;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors.Drawing;
using DevExpress.XtraEditors.ViewInfo;
using DevExpress.Utils.Drawing;

namespace WindowsFormsApplication2
{
    public class MyVistaDataEditCalendar : VistaDateEditCalendar
    {
        public MyVistaDataEditCalendar(RepositoryItemDateEdit item, object editDate)
            : base(item, editDate)
        {
        }

        protected override DevExpress.XtraEditors.Drawing.DateEditPainter CreatePainter()
        {
            return new MyVistaDateEditPainter(this);
        }

        protected override DevExpress.XtraEditors.ViewInfo.DateEditInfoArgs CreateInfoArgs()
        {
            return new MyVistaDateEditInfoArgs(calendar: this);
        }

        internal new Rectangle CaptionRect
        {
            get { return base.CaptionRect; }
        }

        protected override void OnPressedObjectChanged(DevExpress.XtraEditors.Calendar.CalendarHitInfo prev, DevExpress.XtraEditors.Calendar.CalendarHitInfo curr)
        {
            base.OnPressedObjectChanged(prev, curr);
            MyVistaDateEditInfoArgs vdi = Calendars[0] as MyVistaDateEditInfoArgs;
            MyCalendarHitInfo hitInfo = curr as MyCalendarHitInfo;
            Invalidate(prev.Bounds);
            if (curr.HitObject != null)
            {
                if (hitInfo.InLeftYearButton)
                    vdi.LeftYearArrowInfo.State = ObjectState.Pressed;
                if(hitInfo.InRightYearButton)
                    vdi.RightYearArrowInfo.State = ObjectState.Pressed;
            }
            if (curr == null) Invalidate(Bounds);
            else Invalidate(curr.Bounds);
        }

        protected override void OnPressedObjectChanging(DevExpress.XtraEditors.Calendar.CalendarHitInfo prev, DevExpress.XtraEditors.Calendar.CalendarHitInfo curr)
        {
            base.OnPressedObjectChanging(prev, curr);
            MyVistaDateEditInfoArgs vdi = Calendars[0] as MyVistaDateEditInfoArgs;
            MyCalendarHitInfo hitInfo = curr as MyCalendarHitInfo;
            if (curr.HitObject != null)
            {
                vdi.LeftYearArrowInfo.State = hitInfo.InLeftYearButton ? ObjectState.Hot : ObjectState.Normal;
                vdi.RightYearArrowInfo.State = hitInfo.InRightYearButton ? ObjectState.Hot : ObjectState.Normal;
            }
        }

        protected override void OnHotObjectChanged(DevExpress.XtraEditors.Calendar.CalendarHitInfo prevInfo, DevExpress.XtraEditors.Calendar.CalendarHitInfo currInfo)
        {
            base.OnHotObjectChanged(prevInfo, currInfo);
            Invalidate(prevInfo.Bounds);
            MyVistaDateEditInfoArgs vdi = Calendars[0] as MyVistaDateEditInfoArgs;
            MyCalendarHitInfo hitInfo = currInfo as MyCalendarHitInfo;
            vdi.OnAnimationChanged(prevInfo, currInfo);
            if (currInfo != null && currInfo.HitObject != null)
            {
                if (hitInfo.InLeftYearButton && vdi.LeftYearArrowInfo.State != ObjectState.Pressed)
                    vdi.LeftYearArrowInfo.State = ObjectState.Hot;
                if (hitInfo.InRightYearButton && vdi.RightYearArrowInfo.State != ObjectState.Pressed)
                    vdi.RightYearArrowInfo.State = ObjectState.Hot;
            }
        }

        protected override void OnHotObjectChanging(DevExpress.XtraEditors.Calendar.CalendarHitInfo prevInfo, DevExpress.XtraEditors.Calendar.CalendarHitInfo currInfo)
        {
            base.OnHotObjectChanging(prevInfo, currInfo);
            MyVistaDateEditInfoArgs vdi = Calendars[0] as MyVistaDateEditInfoArgs;
            vdi.OnAnimationChanging(prevInfo, currInfo);
            vdi.RightYearArrowInfo.State = ObjectState.Normal;
            vdi.LeftYearArrowInfo.State = ObjectState.Normal;
        }

        protected override void ProcessClick(DevExpress.XtraEditors.Calendar.CalendarHitInfo hit)
        {
            base.ProcessClick(hit);
            if (hit.HitObject != null)
            {
                SwitchType = SwitchStateType.SwitchToRight;
                OnStateChanging(View, View, false, true);
                MyVistaDateEditInfoArgs vdi = Calendars[0] as MyVistaDateEditInfoArgs;
                MyCalendarHitInfo hitInfo = hit as MyCalendarHitInfo;
                if (hitInfo.InRightYearButton)
                    DateTime = DateTime.AddYears(1);
                if (hitInfo.InLeftYearButton)
                    DateTime = DateTime.AddYears(-1);
                OnStateChanged(View, View, false, true);
            }
        }

        internal new bool PrepareAnimation
        {
            get { return base.PrepareAnimation; }
            set { base.PrepareAnimation = value; }
        }
    }
}
